k2/csrc/thread_pool.cu

```diff
@@ -0,0 +1,94 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <utility>
+
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+static int32_t GetDefaultNumThreads() {
+  int num_threads = std::thread::hardware_concurrency();
+  if (num_threads == 0) num_threads = 1;
+  return num_threads;
+}
+
+ThreadPool::ThreadPool(int32_t num_threads)
+    : threads_(num_threads > 0 ? num_threads : GetDefaultNumThreads()) {
+  for (auto &thread : threads_)
+    thread = std::thread([this]() { this->ProcessTasks(); });
+}
+
+ThreadPool::~ThreadPool() {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    keep_running_ = false;
+    not_empty_cond_.notify_all();  // wake up all threads in the pool
+  }
+  for (auto &thread : threads_) thread.join();
```

Comment:
check `if(thread.joinable()` before join.

Reply:
I think `join` is fine here as there is no interface to detach a thread in the pool.